File: evals/relativebench/metrics.py
"""Dependency-free reference implementation of the Protocol v0.1 metrics."""

from collections import Counter
from math import isfinite

VALID_RATINGS = {-2, -1, 0, 1, 2}


def finite_number(value):
    return isinstance(value, (int, float)) and not isinstance(value, bool) and isfinite(value)


def summarize_experience(judgments):
    """Calculate weighted Experience Delta and its published companions."""
    if not judgments:
        raise ValueError("At least one experience judgment is required.")

    total_weight = 0.0
    weighted_rating = 0.0
    better_weight = 0.0
    worse_weight = 0.0
    noticeable_weight = 0.0
    strong_regression_weight = 0.0
    distribution = Counter({rating: 0 for rating in VALID_RATINGS})
    weighted_distribution = Counter({rating: 0 for rating in VALID_RATINGS})

    for judgment in judgments:
        rating = judgment["rating"]
        weight = judgment.get("weight", 1)
        if not finite_number(rating) or rating not in VALID_RATINGS:
            raise ValueError("Experience ratings must be integers from -2 to 2.")
        if not finite_number(weight) or weight <= 0:
            raise ValueError("Judgment weights must be finite and greater than zero.")

        total_weight += weight
        weighted_rating += weight * rating
        distribution[rating] += 1
        weighted_distribution[rating] += weight
        if rating > 0:
            better_weight += weight
        if rating < 0:
            worse_weight += weight
        if rating != 0:
            noticeable_weight += weight
        if rating == -2:
            strong_regression_weight += weight

    if not isfinite(total_weight) or not isfinite(weighted_rating):
        raise ValueError("Accumulated judgment weights exceed the finite numeric range.")
    return {
        "experience_delta": 50 * (weighted_rating / total_weight),
        "preference_lift": 100 * ((better_weight - worse_weight) / total_weight),
        "noticeability": noticeable_weight / total_weight,
        "strong_regression_rate": strong_regression_weight / total_weight,
        "distribution": {str(rating): distribution[rating] for rating in sorted(VALID_RATINGS)},
        "weighted_distribution": {
            str(rating): weighted_distribution[rating] / total_weight for rating in sorted(VALID_RATINGS)
        },
        "sample_size": len(judgments),
        "total_weight": total_weight,
    }


def summarize_flips(pairs):
    """Return the four-cell compatibility matrix and conditional flip rates."""
    stable_successes = negative_flips = positive_flips = stable_failures = 0
    for pair in pairs:
        previous_passed = pair["previous_passed"]
        new_passed = pair["new_passed"]
        if type(previous_passed) is not bool or type(new_passed) is not bool:
            raise ValueError("Scored pair outcomes must be booleans.")
        if previous_passed and new_passed:
            stable_successes += 1
        elif previous_passed:
            negative_flips += 1
        elif new_passed:
            positive_flips += 1
        else:
            stable_failures += 1

    previous_passes = stable_successes + negative_flips
    previous_failures = positive_flips + stable_failures
    return {
        "stable_successes": stable_successes,
        "negative_flips": negative_flips,
        "positive_flips": positive_flips,
        "stable_failures": stable_failures,
        "negative_flip_rate": None if previous_passes == 0 else negative_flips / previous_passes,
        "positive_flip_rate": None if previous_failures == 0 else positive_flips / previous_failures,
    }


def classify_transition(lower_bound, upper_bound, minimum_important_difference=5):
    """Classify an interval using the preregistered provisional threshold."""
    if not all(finite_number(value) for value in (lower_bound, upper_bound, minimum_important_difference)):
        raise ValueError("Classification bounds and threshold must be finite.")
    if lower_bound < -100 or upper_bound > 100:
        raise ValueError("Experience Delta bounds must lie between -100 and 100.")
    if lower_bound > upper_bound:
        raise ValueError("The lower confidence bound cannot exceed the upper bound.")
    if minimum_important_difference < 0:
        raise ValueError("The minimum important difference cannot be negative.")
    if lower_bound > minimum_important_difference:
        return "upgrade"
    if upper_bound < -minimum_important_difference:
        return "regression"
    if lower_bound >= -minimum_important_difference and upper_bound <= minimum_important_difference:
        return "sidegrade"
    return "inconclusive"

File: lib/metrics.ts
export type ExperienceRating = -2 | -1 | 0 | 1 | 2;

export interface ExperienceJudgment {
  rating: ExperienceRating;
  weight?: number;
}

export interface ExperienceSummary {
  experienceDelta: number;
  preferenceLift: number;
  noticeability: number;
  strongRegressionRate: number;
  distribution: Record<ExperienceRating, number>;
  sampleSize: number;
  totalWeight: number;
}

export interface ScoredPair {
  previousPassed: boolean;
  newPassed: boolean;
}

export interface FlipSummary {
  stableSuccesses: number;
  negativeFlips: number;
  positiveFlips: number;
  stableFailures: number;
  negativeFlipRate: number | null;
  positiveFlipRate: number | null;
}

export type TransitionClassification =
  | 'upgrade'
  | 'regression'
  | 'sidegrade'
  | 'inconclusive';

const RATINGS: ExperienceRating[] = [-2, -1, 0, 1, 2];

function cleanZero(value: number): number {
  return Object.is(value, -0) ? 0 : value;
}

export function summarizeExperience(
  judgments: ExperienceJudgment[],
): ExperienceSummary {
  if (judgments.length === 0) {
    throw new Error('At least one experience judgment is required.');
  }

  const distribution = Object.fromEntries(
    RATINGS.map((rating) => [rating, 0]),
  ) as Record<ExperienceRating, number>;

  let totalWeight = 0;
  let weightedRating = 0;
  let betterWeight = 0;
  let worseWeight = 0;
  let noticeableWeight = 0;
  let strongRegressionWeight = 0;

  for (const judgment of judgments) {
    if (!RATINGS.includes(judgment.rating)) {
      throw new Error(`Invalid experience rating: ${judgment.rating}`);
    }

    const weight = judgment.weight ?? 1;
    if (!Number.isFinite(weight) || weight <= 0) {
      throw new Error('Judgment weights must be finite and greater than zero.');
    }

    totalWeight += weight;
    weightedRating += weight * judgment.rating;
    distribution[judgment.rating] += 1;

    if (judgment.rating > 0) betterWeight += weight;
    if (judgment.rating < 0) worseWeight += weight;
    if (judgment.rating !== 0) noticeableWeight += weight;
    if (judgment.rating === -2) strongRegressionWeight += weight;
  }

  return {
    experienceDelta: cleanZero((50 * weightedRating) / totalWeight),
    preferenceLift: cleanZero(
      (100 * (betterWeight - worseWeight)) / totalWeight,
    ),
    noticeability: noticeableWeight / totalWeight,
    strongRegressionRate: strongRegressionWeight / totalWeight,
    distribution,
    sampleSize: judgments.length,
    totalWeight,
  };
}

export function summarizeFlips(pairs: ScoredPair[]): FlipSummary {
  let stableSuccesses = 0;
  let negativeFlips = 0;
  let positiveFlips = 0;
  let stableFailures = 0;

  for (const pair of pairs) {
    if (pair.previousPassed && pair.newPassed) stableSuccesses += 1;
    else if (pair.previousPassed) negativeFlips += 1;
    else if (pair.newPassed) positiveFlips += 1;
    else stableFailures += 1;
  }

  const previousPasses = stableSuccesses + negativeFlips;
  const previousFailures = positiveFlips + stableFailures;

  return {
    stableSuccesses,
    negativeFlips,
    positiveFlips,
    stableFailures,
    negativeFlipRate:
      previousPasses === 0 ? null : negativeFlips / previousPasses,
    positiveFlipRate:
      previousFailures === 0 ? null : positiveFlips / previousFailures,
  };
}

export function classifyTransition(
  lowerBound: number,
  upperBound: number,
  minimumImportantDifference = 5,
): TransitionClassification {
  if (lowerBound > upperBound) {
    throw new Error('The lower confidence bound cannot exceed the upper bound.');
  }
  if (minimumImportantDifference < 0) {
    throw new Error('The minimum important difference cannot be negative.');
  }

  if (lowerBound > minimumImportantDifference) return 'upgrade';
  if (upperBound < -minimumImportantDifference) return 'regression';
  if (
    lowerBound >= -minimumImportantDifference &&
    upperBound <= minimumImportantDifference
  ) {
    return 'sidegrade';
  }
  return 'inconclusive';
}
